Wikipedia/Code/URL+LinkParsing.swift

```diff
@@ -30,17 +30,18 @@ extension URL {
     /// Resolves a relative href from a wiki page against the callee.
     /// The callee should be a standardized page URL generated with wmf_databaseURL, non-article namespaces are OK
     public func resolvingRelativeWikiHref(_ href: String) -> URL? {
-        let urlComponentsString: String
-
-        /// The link is sometimes encoded, and sometimes unencoded. (In some cases, this depends on whether an editor put added an escaped or unescaped version of the URL.) So we remove any potential encoding, so that we can be assured we are starting with an unencoded string.
-        let hrefWithoutEncoding = href.removingPercentEncoding ?? href
-
-        if hrefWithoutEncoding.hasPrefix(".") || hrefWithoutEncoding.hasPrefix("/") {
-            urlComponentsString = hrefWithoutEncoding.addingPercentEncoding(withAllowedCharacters: .relativePathAndFragmentAllowed) ?? href
-        } else {
-            urlComponentsString = hrefWithoutEncoding
+        // The link is sometimes encoded, and sometimes unencoded. (In some cases, this depends on
+        // whether an editor put an escaped or unescaped version of the URL). We percent-encode
+        // non-ASCII characters first because URLComponents cannot ingest strings with
+        // international chacters in some versions of iOS (e.g. 15.5).
+        let urlComponentsString = href.addingPercentEncoding(withAllowedCharacters: .ascii) ?? href
+        var components = URLComponents(string: urlComponentsString)
+        
+        // Encode the URL if it is relative
+        if href.hasPrefix(".") || href.hasPrefix("/") {
+            components?.encodeByComponent() // handle unsafe/disallowed characters
         }
```

Comment:
Thank you for the review! I agree that `encodeByComponent` doesn't seem very necessary here. I don't think it's exactly that `URLComponents(string:)` automatically encodes, but that encoding the non-ASCII characters beforehand is sufficient for the current unit tests since they involve international characters.

I added `encodeByComponent` because the existing code seemed to reflect an intent to specifically encode relative URLs, and I couldn't figure out why from the [commit history](https://github.com/wikimedia/wikipedia-ios/pull/3432/commits/4a15b89175deee8d07811bc4a7aed6dcac62271d) so I wanted to be safe. But like you said it isn't necessary to pass the tests, and intuitively a link that has to be encoded to open on web view would also probably have to be encoded to open on a desktop browser, at which point the link itself should be edited.

While it wouldn't hurt functionality to leave it in, it wouldn't really be good practice to do so if a test doesn't require it so I'm open to removing it if you think that might be better.

Reply:
Sounds good, I think go ahead and remove this `encodeByComponent` part to keep things more simple. We'll keep an eye out for any regressions and will bring it back with a URL test if we find anything. Thanks!